crates/ruff_python_parser/src/semantic_errors.rs

```diff
@@ -1280,6 +1314,143 @@ impl<'a, Ctx: SemanticSyntaxContext> MatchPatternVisitor<'a, Ctx> {
     }
 }
 
+struct Ident {
+    range: TextRange,
+    scope: u32,
+}
+
+struct TryExceptVisitor<'a, Ctx> {
+    /// Context used for emitting errors.
+    ctx: &'a Ctx,
+    identifiers: FxHashMap<ast::name::Name, Ident>,
+    scope_depth: u32,
+}
+
+impl<'a, Ctx> TryExceptVisitor<'a, Ctx> {
+    fn new(ctx: &'a Ctx) -> Self {
+        Self {
+            ctx,
+            identifiers: FxHashMap::default(),
+            scope_depth: 0,
+        }
+    }
+
+    fn add_identifier(&mut self, name: ast::name::Name, range: TextRange) {
+        self.identifiers.insert(
+            name,
+            Ident {
+                range,
+                scope: self.scope_depth,
+            },
+        );
+    }
+}
+
+impl<Ctx> SourceOrderVisitor<'_> for TryExceptVisitor<'_, Ctx>
+where
+    Ctx: SemanticSyntaxContext,
+{
+    fn visit_identifier(&mut self, identifier: &ast::Identifier) {
+        self.add_identifier(identifier.id.clone(), identifier.range);
+    }
+
+    fn visit_expr(&mut self, expr: &Expr) {
+        if let Expr::Name(ast::ExprName { range, id, ctx: _ }) = expr {
```

Comment:
What about lambda expressions?

Reply:
Yes, the visitor traverses into lambdas. The question is whether it should (e.g. we don't traverse into nested functions or classes)